docs/settings.md
<!-- TODO: check matterbridge.toml.sample and see whether this is accurate. ive seen some settings from there undocumented for various protocols -->
# Settings

On this page you will find information about global matterbridge settings. Most of these settings can be applied to specific gateways, and those marked **GENERAL** can be placed in the `[general]` configuration dictionary for setting it across all gateways.

> [!TIP]
> Settings specific to a certain protocol can be found in the corresponding folder in the
> [protocols/](protocols/) documentation folder. API settings can be found in the
> [api/](api/) documentation folder.

# Info

* **OPTIONAL:** this setting isn't enabled by default.
* **RELOADABLE:** this setting can be reloaded by editing the configuration. No restart of matterbridge is required.
* **ALL:** this setting is usable with all bridges.
* **GENERAL:** this setting can be also set under `[general]` which means it's active for all bridges.

# Shared
Only settings which have the `ALL` setting are usable for all bridges.

## EditDisable
Disable sending of edits to other bridges

Setting: OPTIONAL, RELOADABLE \
Format: boolean \
Example: enable it

`EditDisable=true`

## EditSuffix
Message to be appended to every edited message

Setting: OPTIONAL, RELOADABLE \
Format: string \
Example: 

`EditSuffix=" (edited)"`

## EditMaxDays
Edits of messages older than this number of days are ignored. The default of 0 disables this.
So far only works for discord.

Setting: OPTIONAL, RELOADABLE \
Format: int \
Example: ignore edits of messages older than 2 weeks

`EditMaxDays=14`

## IgnoreMessages
Messages you want to ignore.\
Messages matching these regex will be ignored and not sent to other bridges.\
See https://regex-golang.appspot.com/assets/html/index.html for more regex info.

Setting: OPTIONAL, RELOADABLE, ALL \
Format: space seperated regex string \
Example: ignores messages starting with ~~ or messages containing badword

`IgnoreMessages="^~~ badword"`

## IgnoreNicks
Nicks you want to ignore.\
Messages from those users will not be sent to other bridges.

Setting: OPTIONAL, RELOADABLE, ALL \
Format: space seperated regex string \
Example: ignore messages from ircspammer1 and ircspammer2

`IgnoreNicks="ircspammer1 ircspammer2"`

## Label
Extra label that can be used in the `RemoteNickFormat`

Setting: OPTIONAL, RELOADABLE, ALL \
Format: string \
Example: add the label `mychat`

`Label="mychat"`

## PrefixMessagesWithNick
Whether to prefix messages from other bridges with the sender's nick.
Useful if username overrides for incoming webhooks isn't enabled.
 If you set PrefixMessagesWithNick to true, each message
from other bridges will by default be prefixed by "bridge-" + nick. You can,
however, modify how the messages appear, by setting (and modifying) `RemoteNickFormat`

Setting: OPTIONAL, RELOADABLE \
Format: boolean \
Example: 

`PrefixMessagesWithNick=true`


## RemoteNickFormat 
Defines how remote users appear on this bridge. \
The string "{NICK}" (case sensitive) will be replaced by the actual nick / username. \
The string "{BRIDGE}" (case sensitive) will be replaced by the sending bridge. \
The string "{LABEL}" (case sensitive) will be replaced by label= field of the sending bridge. \
The string "{PROTOCOL}" (case sensitive) will be replaced by the protocol used by the bridge. \
The string "{GATEWAY}" (case sensitive) will be replaced by the origin gateway name that is replicating the message. \
The string "{CHANNEL}" (case sensitive) will be replaced by the origin channel name used by the bridge. \
The string "{TENGO}" (case sensitive) will be replaced by the output of the RemoteNickFormat script under `[tengo]` \
The string "{NOPINGNICK}" (case sensitive) will be replaced by the actual nick / username, but with a ZWSP inside the nick, so the irc user with the same nick won't get pinged. See https://github.com/42wim/matterbridge/issues/175 for more information

Setting: OPTIONAL, RELOADABLE, GENERAL, ALL \
Format: string \
Example: add PROTOCOL and NICK

`RemoteNickFormat="[{PROTOCOL}] <{NICK}> "`

## ReplaceMessages
Messages you want to replace. \
It replaces outgoing messages from the bridge. \
So you need to place it by the sending bridge definition.\
Regular expressions supported.

Setting: OPTIONAL, RELOADABLE, ALL \
Format: [ ["from1","to1"],["from2","to2"] ] \
Example: this replaces cat => dog and sleep => awake

`ReplaceMessages=[ ["cat","dog"], ["sleep","awake"] ]`

## ReplaceNicks
Nicks you want to replace. \
See `ReplaceMessages` for the syntax

Setting: OPTIONAL, RELOADABLE, ALL \
Format: [ ["from1","to1"],["from2","to2"] ] \
Example: this replaces user-- => user

`ReplaceNicks=[ ["user--","user"] ]`

## ShowJoinPart
Enable to show users joins/parts from other bridges \
Currently works for messages from the following bridges: irc, mattermost, slack

Setting: OPTIONAL, RELOADABLE, ALL \
Format: boolean \
Example: enable it

`ShowJoinPart=true`

## ShowTopicChange
Enable to show topic changes from other bridges. \
Only works hiding/show topic changes from slack bridge for now. 

Setting: OPTIONAL, RELOADABLE, ALL \
Format: boolean \
Example: enable it

`ShowTopicChange=true`

## SkipTLSVerify
Enable to not verify the certificate on your server.
e.g. when using selfsigned certificates

Setting: OPTIONAL \
Format: boolean \
Example: enable it

`SkipTLSVerify=true`

## StripNick
StripNick only allows alphanumerical nicks. See https://github.com/42wim/matterbridge/issues/285
It will strip other characters from the nick

Setting: OPTIONAL, RELOADABLE, GENERAL, ALL \
Format: boolean \
Example: enable it

`StripNick=true`

## UseLocalAvatar

UseLocalAvatar specifies source bridges for which an avatar should be 'guessed' when an incoming message has no avatar. This works by comparing the username of the message to an existing Discord user, and using the avatar of the Discord user. (Substitute "Discord" with another platform, if used on another platform.)

On Discord, this only works if `WebhookURL` is set (AND the message has no avatar).

At the moment, this setting is only available for Discord.

Setting: OPTIONAL, RELOADABLE  
Format: `["gateway.account1", "gateway.account2", "gateway"]`  
Example: `["irc"]` - this example will guess the avatar coming from the `irc` platform  

# General

Configuration that can be set under `[general]`

## IgnoreFailureOnStart 
Allows you to ignore failing bridges on startup. 
Matterbridge will disable the failed bridge and continue with the other ones. \
Context: https://github.com/42wim/matterbridge/issues/455

Setting: OPTIONAL, RELOADABLE, GENERAL \
Format: boolean \
Example: enable it

`IgnoreFailureOnStart=true`

## LogFile

LogFile defines the location of a file to write logs into, rather than stdout.
Logging will still happen on stdout if the file cannot be opened for #writing, or if the value is empty. 
Note that the log won't roll, so you might want to use logrotate(8) with this feature.

Setting: OPTIONAL, GENERAL \
Format: string \
Example:

`LogFile="/var/log/matterbridge.log"`

## MediaDownloadBlacklist 
Allows you to blacklist specific files from being downloaded.
Filenames matching these regex will not be download/uploaded to the mediaserver. \
You can use regex for this, see https://regex-golang.appspot.com/assets/html/index.html for more regex info

Setting: OPTIONAL, RELOADABLE, GENERAL \
Format: string array \
Example: do not upload html and htm extension

`MediaDownloadBlacklist=[".html$",".htm$"]`

## MediaDownloadPath
MediaDownloadPath is the filesystem path where the media file will be placed, instead of uploaded, if Matterbridge has write access to the directory your webserver is serving. [More information](https://github.com/matterbridge-org/matterbridge/blob/master/docs/advanced/mediaserver.md)

Setting: OPTIONAL, RELOADABLE, GENERAL \
Format: string \
Example: 

`MediaDownloadPath="/srv/http/yourserver.com/public/download"`

## MediaDownloadSize
Maximum size in bytes matterbridge will download for use with upload to the
media server or to other bridges.

Setting: OPTIONAL, RELOADABLE, GENERAL \
Format: int \
Default: 1000000 (1 megabyte) \
Example:

`MediaDownloadSize=1000000`


## MediaServerDownload
The MediaServerDownload will be used so that bridges without native uploading support:
irc and xmpp will be shown links to the files on MediaServerDownload
More information https://github.com/42wim/matterbridge/wiki/Mediaserver-setup-%28advanced%29

Setting: OPTIONAL, RELOADABLE, GENERAL \
Format: string \
Example: 

`MediaServerDownload="https://youserver.com/download"`

## http_proxy

Defines a matterbridge-wide HTTP proxy setting used when making HTTP requests. This is used when:

- downloading remote HTTP attachments
- connecting to HTTP APIs for protocols that support it:
  - discord

This setting is not used when connecting to the media server to upload attachments.

- Setting: **OPTIONAL**
- Format: *string*
- Example:
  ```toml
  http_proxy="http://login:password@server.example.org:1234"
  ```


changelog.md
# unreleased

## BREAKING

- Main development repository is now https://github.com/matterbridge-org/matterbridge
- History was rewritten to remove the vendor/ folder, easing PR reviews and making the
  repository much lighter. [See issue community/#5](https://github.com/matterbridge-org/community/issues/5).
- Removed protocols, see [issue community/#9](https://github.com/matterbridge-org/matterbridge/issues/9)
  - harmony protocol does not exist anymore
  - gitter protocol does not exist anymore ; gitter.im is now a matrix protocol server
  - steam protocol has changed profoundly
  - keybase has been removed because we don't have a maintainer for it. See
    [issue #9](https://github.com/matterbridge-org/matterbridge/issues/9)
- During attachment filename sanitation, extensions are also sanitized
- Attachment filenames are truncated to 50 characters, and discarded entirely
  when extension (anything after the first `.`) is longer than that
- matrix: Change to mautrix.go for the matrix backend. See ([pr #79](https://github.com/matterbridge-org/matterbridge/pull/79)/[issue #60](https://github.com/matterbridge-org/matterbridge/issues/60)
- xmpp: Initial replies/edits support has been removed, because it was incorrect ([#12](https://github.com/matterbridge-org/matterbridge/pull/12))
- xmpp: `NoTls` setting has been deprecated; to disable `StartTls` and start a plaintext connection, use `NoStartTls`
- xmpp: `WebhookURL` for slack-compatible webhooks has been removed (see XMPP docs) ; you can
  open an issue if you'd like to use it again and can document the setup
- Go required version is now v1.26
- `MediaServerUpload` has been deprecated ([#30](https://github.com/matterbridge-org/matterbridge/issues/30)),
  because we don't know how to make it work and were improving those parts of the codebase.
  If you were successfully using it, your feedback is welcome and we may reintroduce it.
- irc: Leading colon messages are no longer doubled by default as an undocumented hack (eg `:D` -> `::D`); it's now enabled by the `DoubleColonPrefix` setting.
  If you are using this setting please help us understand the usecase by commenting
  on [issue #122](https://github.com/matterbridge-org/matterbridge/issues/122), otherwise this setting may be deprecated in the near-future.
- irc: Charset for irc bridges will now be set to a default of "UTF-8", to avoid mojibake when attempting to automatically guess the incoming charset.  If you want to try autodetecting, you will need to set this to "autodetect" for the irc bridge.  Ideally, you will know which charset to set and won't have to try to guess.  Even with autodetection set, UTF-8 will be checked first before trying anything else, then fall back to latin-1 if the autodetection fails.  This should mostly address ([#120](https://github.com/matterbridge-org/matterbridge/issues/120))
- irc: Destination bridges which have `Colornicks` set, and are not using `StripNick` nor `UseRelayMsg`, when receiving from bridges that allow spaces in the nickname (e.g. Discord, XMPP), will see those spaces in the nick replaced with non-breaking spaces.  This is to facilitate using spaces as delimiters for the new `Colornicks` behavior ([#218](https://github.com/matterbridge-org/matterbridge/pull/218))
- irc: Messages are now split by default at word boundaries, with a maximum length per line of `MessageLength` (default of 512), and which now takes into account the channel name, prefix length (botnick!user@host), formatted remote nick, and other padding.  This is now controlled by the `MessageSplit` setting, which when disabled, will rely on the `girc` library to perform any required splitting.  The `MessageLength` setting can be overridden by a server that sends a LINELEN token in an ISUPPORT message when connecting.  Fixes ([#190](https://github.com/matterbridge-org/matterbridge/issues/190))

## New Features

- general
  - matterbridge output now colors log level for easier log reading ([#25](https://github.com/matterbridge-org/matterbridge/pull/25))
  - new HTTP helpers are common to all bridges, and allow overriding specific settings ([#59](https://github.com/matterbridge-org/matterbridge/pull/59))
  - matterbridge is now built with whatsappmulti backend enabled by default, unless the `nowhatsappmulti` build tag is passed
  - Docker images are now automatically built and published to `ghcr.io/matterbridge-org/matterbridge` ([#86](https://github.com/matterbridge-org/matterbridge/pull/86))
  - matterbridge will now apply a default `RemoteNickFormat` setting of `"[{PROTOCOL}] <{NICK}> "` which may be overridden by individual bridge settings, environment variables, or the `General` section of the config file, fulfilling the enhancement requested at ([#162](https://github.com/matterbridge-org/matterbridge/issues/162))
- matrix
  - Supports MSC4144/puppeting ([#232](https://github.com/matterbridge-org/matterbridge/pulls/232)). See also [MSC4144](https://github.com/matrix-org/matrix-spec-proposals/pulls/4144). Note that this is useless unless you have a client that can display these. Clients that don't will fall back to displaying e.g. `Nick: msg`.
  - the Viper configuration functions have been updated to defer a panic-handling function instead of deferring their RWMutex RUnlock calls.  This became necessary due to the new "SetVal" function, which may be used to override a configuration setting; this is now the first time a write lock has been used within the config package.  Otherwise, obtaining a write lock could have caused matterbridge to behave as a single-threaded application, due to the numerous RLock calls made from multiple bridges during runtime.

bridge/config/config.go
package config

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify" // TODO: lock viper for writing when the config file changes, as that is not concurrency safe
	"github.com/sirupsen/logrus"
	"github.com/spf13/viper"
)

const (
	EventJoinLeave         = "join_leave" // left for backwards compatibility
	EventJoin              = "join"
	EventLeave             = "leave"
	EventTopicChange       = "topic_change"
	EventFailure           = "failure"
	EventFileFailureSize   = "file_failure_size"
	EventAvatarDownload    = "avatar_download"
	EventRejoinChannels    = "rejoin_channels"
	EventUserAction        = "user_action"
	EventMsgDelete         = "msg_delete"
	EventFileDelete        = "file_delete"
	EventAPIConnected      = "api_connected"
	EventUserTyping        = "user_typing"
	EventGetChannelMembers = "get_channel_members"
	EventNoticeIRC         = "notice_irc"
)

const ParentIDNotFound = "msg-parent-not-found"

type Message struct {
	Text      string    `json:"text"`
	Channel   string    `json:"channel"`
	Username  string    `json:"username"`
	UserID    string    `json:"userid"` // userid on the bridge
	Avatar    string    `json:"avatar"`
	Account   string    `json:"account"`
	Event     string    `json:"event"`
	Protocol  string    `json:"protocol"`
	Gateway   string    `json:"gateway"`
	ParentID  string    `json:"parent_id"`
	Timestamp time.Time `json:"timestamp"`
	ID        string    `json:"id"`
	Extra     map[string][]any
}

func (m Message) ParentNotFound() bool {
	return m.ParentID == ParentIDNotFound
}

func (m Message) ParentValid() bool {
	return m.ParentID != "" && !m.ParentNotFound()
}

// GetFileInfos extracts typed FileInfo list from the message.
//
// This method is guaranteed not to fail. The inner type casting should never
// fail, but will simply produce a warning if that's the case.
func (m Message) GetFileInfos(log *logrus.Entry) *[]*FileInfo {
	var fileInfos []*FileInfo

	for _, file := range m.Extra["file"] {
		fileInfo, ok := file.(FileInfo)
		if !ok {
			// This should never happen, unless a bridge receiving an external message
			// produces an invalid Extra field where the File is not valid FileInfo.
			// TODO: log more information about the message for debugging.
			log.Warn(FileCastError())
			continue
		}

		fileInfos = append(fileInfos, &fileInfo)
	}

	return &fileInfos
}

// FileInfo is an attachment contained in a message.
//
// When receiving an attachment (eg. an image), a bridge should populate the
// Data/Size fields.
//
// When the media server is enabled, for services that don't support file upload
// (such as IRC), the gateway router will upload the file to the media server
// and populate the URL/SHA fields. The Data/Size fields are not removed
// in this process. See handleFiles in gateway/handlers.go
type FileInfo struct {
	Name     string
	Data     *[]byte
	Comment  string
	URL      string
	Size     int64
	Avatar   bool
	SHA      string
	NativeID string
}

var errFileCast = errors.New("failed to cast config.FileInfo")

func FileCastError() error {
	return fmt.Errorf("%w", errFileCast)
}

type ChannelInfo struct {
	Name        string
	Account     string
	Direction   string
	ID          string
	SameChannel map[string]bool
	Options     ChannelOptions
}

type ChannelMember struct {
	Username    string
	Nick        string
	UserID      string
	ChannelID   string
	ChannelName string
}

type ChannelMembers []ChannelMember

type Protocol struct {
	AllowMention           []string // discord
	BindAddress            string   // mattermost, slack // DEPRECATED
	Buffer                 int      // api
	Charset                string   // irc
	ClientID               string   // msteams
	Casemapping            string   // IRC, auto-configured setting for allowable characters in nicks, not configurable
	ColorNicks             bool     // only irc for now
	CustomStatus           string   // discord
	Debug                  bool     // general
	DebugLevel             int      // only for irc now
	DeviceID               string   // matrix
	DisableMarkdownParsing bool     // matrix
	DisableWebPagePreview  bool     // telegram
	EditSuffix             string   // mattermost, slack, discord, telegram
	EditDisable            bool     // mattermost, slack, discord, telegram
	EditMaxDays            int      // discord
	EmbedFormat            string   // discord
	HTMLDisable            bool     // matrix
	IconURL                string   // mattermost, slack
	IgnoreFailureOnStart   bool     // general
	IgnoreNicks            string   // all protocols
	IgnoreMessages         string   // all protocols
	Jid                    string   // xmpp
	JoinDelay              string   // all protocols
	Label                  string   // all protocols
	Login                  string   // mattermost, matrix
	LogFile                string   // general
	MediaDownloadBlackList []string
	MediaDownloadPath      string // Write upload to a file on the same server.
	MediaDownloadSize      int    // all protocols
	MediaServerDownload    string
	MediaConvertTgs        string     // telegram
	MediaConvertWebPToPNG  bool       // telegram
	MessageDelay           int        // IRC, time in millisecond to wait between messages
	MessageFormat          string     // telegram
	MessageLength          int        // IRC, max length of a message allowed, defaults to 512 (counting CRLF)
	MessagePrefix          int        // IRC, current length of message prefix for bot, not configurable
	MessageQueue           int        // IRC, size of message queue for flood control
	MessageSplit           bool       // IRC, split long messages, default true.  If set false, let the irc library handle splitting
	MessageSplitMaxCount   int        // discord, split long messages into at most this many messages instead of clipping (MessageLength=1950 cannot be configured)
	Muc                    string     // xmpp
	MxID                   string     // matrix
	Name                   string     // all protocols
	Nick                   string     // all protocols
	NickFormatter          string     // mattermost, slack
	NickServNick           string     // IRC
	NickServUsername       string     // IRC
	NickServPassword       string     // IRC
	NicksPerRow            int        // mattermost, slack
	NoHomeServerSuffix     bool       // matrix
	NoSendJoinPart         bool       // all protocols
	NoTLS                  bool       // mattermost, xmpp
	Password               string     // IRC,mattermost,XMPP,matrix
	PickleKey              string     // matrix
	PrefixMessagesWithNick bool       // mattemost, slack
	PreserveThreading      bool       // slack
	Protocol               string     // all protocols
	QuoteDisable           bool       // telegram,discord
	QuoteFormat            string     // telegram,discord
	QuoteLengthLimit       int        // telegram,discord
	RealName               string     // IRC
	RecoveryKey            string     // matrix
	RejoinDelay            int        // IRC
	RelayFallbackNick      string     // IRC, fallback nick to use when SanitizeNick results in an empty message
	RelayMsgSep            string     // IRC, autodetected, required separator char(s) in relayed nicks, not configurable
	ReplaceMessages        [][]string // all protocols
	ReplaceNicks           [][]string // all protocols
	RemoteNickFormat       string     // all protocols
	RunCommands            []string   // IRC
	Server                 string     // IRC,mattermost,XMPP,discord,matrix
	SessionFile            string     // msteams,whatsapp
	ShowJoinPart           bool       // all protocols
	ShowTopicChange        bool       // slack
	ShowUserTyping         bool       // slack
	ShowEmbeds             bool       // discord
	SkipEncrypted          bool       // xmpp
	SkipTLSVerify          bool       // IRC, mattermost
	SkipVersionCheck       bool       // mattermost
	StripNick              bool       // all protocols
	StripMarkdown          bool       // irc
	SyncTopic              bool       // slack
	TengoModifyMessage     string     // general
	Team                   string     // mattermost
	TeamID                 string     // msteams
	TenantID               string     // msteams
	Token                  string     // slack, discord, api, matrix, stoat
	Topic                  string     // zulip
	URL                    string     // mattermost, slack // DEPRECATED
	UseAPI                 bool       // mattermost, slack
	UseLocalAvatar         []string   // discord
	UseSASL                bool       // IRC
	UseTLS                 bool       // IRC
	UseDiscriminator       bool       // discord
	UseFirstName           bool       // telegram
	UseUserName            bool       // discord, matrix, mattermost
	UseInsecureURL         bool       // telegram
	UseMSC4144             bool       // matrix
	UserName               string     // IRC
	UseRelayFallback       bool       // IRC, controls whether RelayFallbackNick is used, defaults to true
	UseRelayMsg            bool       // IRC
	VerboseJoinPart        bool       // IRC
	WebhookBindAddress     string     // mattermost, slack
	WebhookURL             string     // mattermost, slack
}

type ChannelOptions struct {
	Key        string // irc, xmpp
	WebhookURL string // discord
	Topic      string // zulip
}

type Bridge struct {
	Account     string
	Channel     string
	Options     ChannelOptions
	SameChannel bool
}

type Gateway struct {
	Name   string
	Enable bool
	In     []Bridge
	Out    []Bridge
	InOut  []Bridge
}

type Tengo struct {
	InMessage        string
	Message          string
	RemoteNickFormat string
	OutMessage       string
}

type SameChannelGateway struct {
	Name     string
	Enable   bool
	Channels []string
	Accounts []string
}

type BridgeValues struct {
	API                map[string]Protocol
	IRC                map[string]Protocol
	Mattermost         map[string]Protocol
	Matrix             map[string]Protocol
	Slack              map[string]Protocol
	SlackLegacy        map[string]Protocol
	Steam              map[string]Protocol
	XMPP               map[string]Protocol
	Discord            map[string]Protocol
	Telegram           map[string]Protocol
	Rocketchat         map[string]Protocol
	SSHChat            map[string]Protocol
	WhatsApp           map[string]Protocol // TODO is this struct used? Search for "SlackLegacy" for example didn't return any results
	Zulip              map[string]Protocol
	Keybase            map[string]Protocol
	Mumble             map[string]Protocol
	Fluxer             map[string]Protocol
	Stoat              map[string]Protocol
	General            Protocol
	Tengo              Tengo
	Gateway            []Gateway
	SameChannelGateway []SameChannelGateway
}

type Config interface {
	Viper() *viper.Viper
	BridgeValues() *BridgeValues
	IsKeySet(key string) bool
	GetBool(key string) (bool, bool)
	GetInt(key string) (int, bool)
	GetString(key string) (string, bool)
	GetStringSlice(key string) ([]string, bool)
	GetStringSlice2D(key string) ([][]string, bool)
	IsFilenameBlacklisted(filename string) bool
	SetVal(key string, value any)
}

type config struct {
	sync.RWMutex

	logger                        *logrus.Entry
	v                             *viper.Viper
	cv                            *BridgeValues
	MediaDownloadBlackListRegexes *[]*regexp.Regexp
}

// NewConfig instantiates a new configuration based on the specified configuration file path.
func NewConfig(rootLogger *logrus.Logger, cfgfile string) Config {
	logger := rootLogger.WithFields(logrus.Fields{"prefix": "config"})

	viper.SetConfigFile(cfgfile)

	input, err := os.ReadFile(cfgfile) //nolint:gosec
	if err != nil {
		logger.Fatalf("Failed to read configuration file: %#v", err)
	}

	cfgtype := detectConfigType(cfgfile)
	mycfg := newConfigFromString(logger, input, cfgtype)
	if mycfg.cv.General.LogFile != "" {
		logfile, err := os.OpenFile(mycfg.cv.General.LogFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0600)
		if err == nil {
			logger.Info("Opening log file ", mycfg.cv.General.LogFile)
			rootLogger.Out = logfile
		} else {
			logger.Warn("Failed to open ", mycfg.cv.General.LogFile)
		}
	}
	if mycfg.cv.General.MediaDownloadSize == 0 {
		mycfg.cv.General.MediaDownloadSize = 1000000
	}

	// Precompile MediaBlackList regexes so we make sure they're correct,
	// and they don't have to be compiled on every file attachment, because
	// that's a slow operation.
	mycfg.compileMediaDownloadBlackListRegexes()

	viper.WatchConfig()
	viper.OnConfigChange(func(e fsnotify.Event) {
		logger.Println("Config file changed:", e.Name)
	})

	return mycfg
}

// NewConfigFromString instantiates a new configuration based on the specified string.
func NewConfigFromString(rootLogger *logrus.Logger, input []byte) Config {
	logger := rootLogger.WithFields(logrus.Fields{"prefix": "config"})
	return newConfigFromString(logger, input, "toml")
}

func (c *config) BridgeValues() *BridgeValues {
	return c.cv
}

func (c *config) Viper() *viper.Viper {
	return c.v
}

func (c *config) IsKeySet(key string) bool {
	defer c.handlePanic()

	c.RLock()
	isset := c.v.IsSet(key)
	c.RUnlock()

	return isset
}

func (c *config) GetBool(key string) (bool, bool) {
	defer c.handlePanic()

	c.RLock()
	mykey := c.v.GetBool(key)
	isset := c.v.IsSet(key)
	c.RUnlock()

	return mykey, isset
}

func (c *config) GetInt(key string) (int, bool) {
	defer c.handlePanic()

	c.RLock()
	mykey := c.v.GetInt(key)
	isset := c.v.IsSet(key)
	c.RUnlock()

	return mykey, isset
}

func (c *config) GetString(key string) (string, bool) {
	defer c.handlePanic()

	c.RLock()
	mykey := c.v.GetString(key)
	isset := c.v.IsSet(key)
	c.RUnlock()

	return mykey, isset
}

func (c *config) GetStringSlice(key string) ([]string, bool) {
	defer c.handlePanic()

	c.RLock()
	mykey := c.v.GetStringSlice(key)
	isset := c.v.IsSet(key)
	c.RUnlock()

	return mykey, isset
}

func (c *config) GetStringSlice2D(key string) ([][]string, bool) {
	defer c.handlePanic()

	var result [][]string

	c.RLock()

	res, ok := c.v.Get(key).([]any)
	if !ok {
		c.RUnlock()

		return nil, false
	}

	for idx, entry := range res {
		entryarray, ok := entry.([]string)
		if !ok {
			c.logger.Errorf("GetStringSlice2D: key %s[%d] should be an array of strings", key, idx)
			return nil, false
		}

		// TODO: This method should only accept 2-entries arrays in the inner arrays
		// TODO: Add tests
		result = append(result, entryarray)
	}

	c.RUnlock()

	return result, true
}

func (c *config) SetVal(key string, value any) {
	defer c.handlePanic()

	c.Lock()
	c.v.Set(key, value)
	c.Unlock()
}

// IsFilenameBlackListed checks if a given file name matches the
// configured blacklist. This is useful to filter potentially-harmful
// files that could be served over HTTP (eg. `.html` with XSS).
func (c *config) IsFilenameBlacklisted(filename string) bool {
	defer c.handlePanic()

	c.RLock()
	for _, re := range *c.MediaDownloadBlackListRegexes {
		if re.MatchString(filename) {
			c.RUnlock()

			return true
		}
	}

	c.RUnlock()

	return false
}

func GetIconURL(msg *Message, iconURL string) string {
	info := strings.Split(msg.Account, ".")
	protocol := info[0]
	name := info[1]
	iconURL = strings.ReplaceAll(iconURL, "{NICK}", msg.Username)
	iconURL = strings.ReplaceAll(iconURL, "{BRIDGE}", name)
	iconURL = strings.ReplaceAll(iconURL, "{PROTOCOL}", protocol)

	return iconURL
}

type TestConfig struct {
	Config

	Overrides map[string]any
}

func (c *TestConfig) IsKeySet(key string) bool {
	_, ok := c.Overrides[key]
	return ok || c.Config.IsKeySet(key)
}

func (c *TestConfig) GetBool(key string) (bool, bool) {
	val, ok := c.Overrides[key]
	if ok {
		return val.(bool), true
	}
	return c.Config.GetBool(key)
}

func (c *TestConfig) GetInt(key string) (int, bool) {
	if val, ok := c.Overrides[key]; ok {
		return val.(int), true
	}
	return c.Config.GetInt(key)
}

func (c *TestConfig) GetString(key string) (string, bool) {
	if val, ok := c.Overrides[key]; ok {
		return val.(string), true
	}
	return c.Config.GetString(key)
}

func (c *TestConfig) GetStringSlice(key string) ([]string, bool) {
	if val, ok := c.Overrides[key]; ok {
		return val.([]string), true
	}
	return c.Config.GetStringSlice(key)
}

func (c *TestConfig) GetStringSlice2D(key string) ([][]string, bool) {
	if val, ok := c.Overrides[key]; ok {
		return val.([][]string), true
	}
	return c.Config.GetStringSlice2D(key)
}

func (c *config) compileMediaDownloadBlackListRegexes() {
	regexes := []*regexp.Regexp{}

	// TODO: apparently c.cv.General does not get updated when config reloads
	// see https://github.com/matterbridge-org/matterbridge/issues/57
	// for _, regex := range c.cv.General.MediaDownloadBlackList {
	for _, regex := range c.v.GetStringSlice("general.MediaDownloadBlackList") {
		c.logger.Debugf("Found blacklist regex %s", regex)

		re, err := regexp.Compile(regex)
		if err != nil {
			c.logger.Errorf("incorrect regexp %s for MediaDownloadBlackList", regex)
			continue
		}

		regexes = append(regexes, re)
	}

	c.MediaDownloadBlackListRegexes = &regexes
	c.logger.Debug("Successfully applied new `MediaDownloadBlackList` regexes")
}

// detectConfigType detects JSON and YAML formats, defaults to TOML.
func detectConfigType(cfgfile string) string {
	fileExt := filepath.Ext(cfgfile)
	switch fileExt {
	case ".json":
		return "json"
	case ".yaml", ".yml":
		return "yaml"
	}
	return "toml"
}

// TODO: Detect whether any locks are currently set (a feature for debug mode only)
func (c *config) handlePanic() {
	r := recover()
	if r != nil {
		c.logger.Warnf("Recovered from panic: %#v", r)
	}
}

func newConfigFromString(logger *logrus.Entry, input []byte, cfgtype string) *config {
	viper.SetConfigType(cfgtype)
	viper.SetDefault("General.RemoteNickFormat", "[{PROTOCOL}] <{NICK}> ") // fixes #162
	viper.SetDefault("General.Charset", "utf-8")                           // fixes #120 (it's irc-only, but shouldn't hurt to put it here)
	viper.SetDefault("General.MessageSplit", true)                         // fixes #190 (irc-only, but should be fine here.  Override it to prefer the girc split function)
	viper.SetEnvPrefix("matterbridge")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_", "-", "_"))
	viper.AutomaticEnv()

	err := viper.ReadConfig(bytes.NewBuffer(input))
	if err != nil {
		logger.Fatalf("Failed to parse the configuration: %s", err)
	}

	cfg := &BridgeValues{}
	err = viper.Unmarshal(cfg)
	if err != nil {
		logger.Fatalf("Failed to load the configuration: %s", err)
	}
	return &config{
		logger: logger,
		v:      viper.GetViper(),
		cv:     cfg,
	}
}
